### PR TITLE
Mock driver plugin implementation and driver registration

### PR DESCRIPTION
**Pull request by nickethier**
Started by @nickethier - finished by @schmichael

This implements the mock driver as a plugin. All plugins are registered by default, but the mock driver is *not* registered for release builds.

I created files specifically for registering plugins to hopefully ease management of build tags (as we have a lot of platform specific drivers).

### Why not use side-effect imports?

Before this PR drivers registered themselves in init methods and the `agent` package imported all drivers to register them.

This PR moves registrations into the `catalog` package and references static variables within the driver packages.

#### Pros of registration in `catalog` package (new approach)

* Registration is near the registrar. Traditionally registration by side effect can be difficult to discover; hopefully putting everything registration related in one place helps.
* Tests outside of the `agent` package would have to manually register the plugins they needed.

#### Cons of registration in `catalog` package

* Drivers cannot import the `catalog` package. Not an issue today, and hopefully never an issue in the future.
* All plugins will be registered by default. May negatively affect test build/runtime performance.

### CODE REVIEW

File: plugins/shared/catalog/testing.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// +build !release
```
Comment (dadgar) [inline plugins/shared/catalog/testing.go]: I don't think this is necessary. It shouldn't be imported by anything in the build, only tests. None of our other testing files do this

File: nomad/testing.go [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// +build !release
```
Comment (dadgar) [inline nomad/testing.go]: Same comment

File: drivers/rawexec/driver.go [hunk 3]
```diff
@@ -17,33 +17,34 @@ import (
...
-			if v, err := strconv.ParseBool(opts["driver.raw_exec.no_cgroups"]); err == nil {
-				conf["no_cgroups"] = v
-			}
-			return conf, nil
-		},
-	)
+	}
+)
+
+func PluginLoader(opts map[string]string) (map[string]interface{}, error) {
```
Comment (dadgar) [inline drivers/rawexec/driver.go]: Put a comment on all three of these public fields

File: drivers/mock/driver.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+	d.logger.Debug("killing task",
+		"task_name", h.task.Name,
+		"kill_after", h.killAfter,
+		"kill_timeout", h.killTimeout,
+	)
+
+	select {
+	case <-h.waitCh:
+		d.logger.Debug("not killing task: already exited", "task_name", h.task.Name)
+	case <-time.After(h.killAfter):
```
Comment (dadgar) [inline drivers/mock/driver.go]: This doesn't make sense. It is waiting on unrelated timeouts.

Comment (schmichael) [reply drivers/mock/driver.go]: Right? I just ported the existing code and it confused me too: https://github.com/hashicorp/nomad/blob/master/client/driver/mock_driver.go#L380-L402

Should I do a grep and switch everything to `kill_timeout` and remove `kill_after`?

Comment (schmichael) [reply drivers/mock/driver.go]: @nickethier pointed out killTimeout isn't used. Removing that one.

Comment (dadgar) [reply drivers/mock/driver.go]: Ah I think I see. If not set the kill will happen immediately. So it is just a way to introduce a kill delay.

**Pull request merged**